### PR TITLE
Adjust Market Order Logic When Limit Orders Are Disabled

### PR DESCRIPTION
**Pull request by nlordell**
This PR adjusts the order classification logic.

Specifically, when limit orders were disabled, it was possible to create market orders with a 0-fee (as the 0-fee would cause it to skip the `get_quote_and_check_fee` fee check but disabled limit orders make the order be considered a market order).

@fedgiac and I already applied a hotfix in production, this PR adds the fix to `main` as 
well. This is not urgent (as limit orders are enabled in staging, so it is not possible to create 0-fee market orders ATM).

Additionally, this PR also:
- restores the behaviour of out-of-market orders being flagged as liquidity orders
- only stores quotes at order creation if an existing quote wasn't found. This is to avoid orders that are created with a specific quote ID getting a different one.

### Test Plan

Changed the test that checks 0-fee market orders error validation when limit orders are disabled, and added a new test to prevent regressions of the liquidity classification for out-of market orders.

### CODE REVIEW

File: crates/shared/src/order_validation.rs [hunk 1]
```diff
@@ -379,15 +379,23 @@ impl OrderValidating for OrderValidator {
             return Err(ValidationError::ZeroAmount);
         }
 
-        let liquidity_owner = self.liquidity_order_owners.contains(&owner);
+        let class = if self.liquidity_order_owners.contains(&owner) {
+            OrderClass::Liquidity
+        } else if self.enable_limit_orders && order.data.fee_amount.is_zero() {
+            OrderClass::Limit
```
Comment (MartinquaXD) [inline crates/shared/src/order_validation.rs:385]: Couldn't we just return `Liquidity` if `!self.enable_limit_orders && order.data.fee_amount.is_zero()`?
Then we wouldn't have to reclassify this particular order and the code becomes easier to understand. Or am I missing some side effect we need?
Also wouldn't we still allow market orders with 0 fee if they are within market price with this PR?

Comment (nlordell) [reply crates/shared/src/order_validation.rs:385]: The reclassification is needed for market orders that are out of price. We don't have that information at this point in the code.

File: crates/shared/src/order_validation.rs [hunk 2]
```diff
@@ -398,7 +406,7 @@ impl OrderValidating for OrderValidator {
             from: owner,
             app_data: order.data.app_data,
         };
-        let quote = if !liquidity_owner && order.data.fee_amount > U256::zero() {
+        let quote = if class == OrderClass::Market {
```
Comment (nlordell) [inline crates/shared/src/order_validation.rs:409]: @MartinquaXD This specifically is the change that fixes 0-fee market orders.

**Review by MartinquaXD — APPROVED** [review]

**Comment by nlordell** [comment]
Merging this with a single approval as @fedgiac reviewed the changes with me when creating the v2.80.2 release and this is an important change for next week.

**Pull request merged**